File: LeetCode - Top Interview 150/21.py
# 21. Merge Two Sorted Lists
# https://leetcode.com/problems/merge-two-sorted-lists/


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        list1_ptr = list1
        list2_ptr = list2

        dummy = current = ListNode()

        while list1_ptr or list2_ptr:
            if list1_ptr and not list2_ptr:
                current.next = list1_ptr
                list1_ptr = list1_ptr.next

            elif not list1_ptr and list2_ptr:
                current.next = list2_ptr
                list2_ptr = list2_ptr.next

            elif list1_ptr.val < list2_ptr.val:
                current.next = list1_ptr
                list1_ptr = list1_ptr.next

            else:
                current.next = list2_ptr
                list2_ptr = list2_ptr.next

            current = current.next

        return dummy.next

File: LeetCode - Algorithm/Linked List/21.ts
// 21. Merge Two Sorted Lists
// https://leetcode.com/problems/merge-two-sorted-lists/

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  const dummy = new ListNode(-1);
  let current = dummy;

  let first = list1;
  let second = list2;

  while (first || second) {
    if (!first && second) {
      current.next = second;
      current = current.next;
      second = second.next;
    } else if (first && !second) {
      current.next = first;
      current = current.next;
      first = first.next;
    } else if (first!.val >= second!.val) {
      current.next = second!;
      current = current.next;
      second = second!.next;
    } else {
      current.next = first!;
      current = current.next;
      first = first!.next;
    }
  }

  return dummy.next;
}
